Initial implementation of FakeTensor caching

masnesral opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113873

Summary: Cache the result of FakeTensor dispatch and skip re-evaluation on cache hits.

Test Plan: New unit tests. Caching is enabled in this diff, so all existing tests exercise the cache as well.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

Differential Revision: [D52841637](https://our.internmc.facebook.com/intern/diff/D52841637)

masnesral committed:
[RFC] Initial implementation of FakeTensor caching

Summary: Cache the result of FakeTensor dispatch and skip re-evaluation on cache hits.

Test Plan:

[ghstack-poisoned]

masnesral committed:
Update on "[RFC] Initial implementation of FakeTensor caching"


Summary: Cache the result of FakeTensor dispatch and skip re-evaluation on cache hits.

Test Plan:

[ghstack-poisoned]

masnesral committed:
Update on "[RFC] Initial implementation of FakeTensor caching"


Summary: Cache the result of FakeTensor dispatch and skip re-evaluation on cache hits.

Test Plan:

[ghstack-poisoned]

masnesral committed:
Update on "[RFC] Initial implementation of FakeTensor caching"


Summary: Cache the result of FakeTensor dispatch and skip re-evaluation on cache hits.

Test Plan:

[ghstack-poisoned]

masnesral committed:
Update on "[RFC] Initial implementation of FakeTensor caching"


Summary: Cache the result of FakeTensor dispatch and skip re-evaluation on cache hits.

Test Plan:

[ghstack-poisoned]

masnesral committed:
Update on "[RFC] Initial implementation of FakeTensor caching"


Summary: Cache the result of FakeTensor dispatch and skip re-evaluation on cache hits.

Test Plan:

[ghstack-poisoned]

masnesral committed:
Update on "[RFC] Initial implementation of FakeTensor caching"


Summary: Cache the result of FakeTensor dispatch and skip re-evaluation on cache hits.

Test Plan:

[ghstack-poisoned]

masnesral committed:
Update on "[RFC] Initial implementation of FakeTensor caching"


Summary: Cache the result of FakeTensor dispatch and skip re-evaluation on cache hits.

Test Plan:

[ghstack-poisoned]

masnesral committed:
Update on "[RFC] Initial implementation of FakeTensor caching"


Summary: Cache the result of FakeTensor dispatch and skip re-evaluation on cache hits.

Test Plan:

[ghstack-poisoned]

masnesral commented:
@eellison / @bdhirsh / @Chillee: This is probably ready for a first round of eyeballs at your convenience. It's missing any specific unit tests. I'll get cranking on those if ppl don't find any red flags in the RFC. This PR enables caching by default, so I was able to work through a lot of failures via all the existing tests.

I'm overall vaguely concerned about brittleness. I'll make some comments inline, but TLDR it took me a while to sort through failures in existing unit tests that were due to, e.g., weird ops that behave in a way that make them break caching, or "mode changes" that don't appear in the arguments to `dispatch` but can change the behavior of some op. I just don't know what other corner cases exist out there or global knobs that can change behavior... Lemme know your thoughts

masnesral commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1306,6 +1309,107 @@ def tolist(self):
     __torch_function__ = torch._C._disabled_torch_function_impl
 
 
+@dataclass(frozen=True)
```
This class is essentially duplicated with another that I added in the previous code cache work. I'll combine them, just lemme know where they belong conceptually.Is there an appropriate "utils" module, for example? Or maybe it's fine if they go here.

masnesral commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1414,6 +1524,215 @@ def __exit__(self, a, b, c):
+        # Avoid caching for any ops that would require a more sophisticated
+        # caching implementation, e.g., data dependent ops or ops that modify
+        # the inputs.
+        if torch.Tag.data_dependent_output in func.tags:
```
This is the random list of "if special-case-x then don't cache" cases I found. It makes me worried that I won't have found every possibility. It's going to be quite confusing for users if we ever serve a cache hit that would actually produce a different result if actually evaluated.

eellison replied:
Yea, that is concerning. We should have a mode of running comparing cached results with a re-run, and see if the metadata differs. 

cc @bdhirsh @ezyang @voznesenskym  for thoughts

masnesral commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1414,6 +1524,215 @@ def __exit__(self, a, b, c):
+            # Capture the default_dtype mode since that can affect the output tensor,
+            # e.g., when operating on constant float values.
+            torch.get_default_dtype(),
+            # The disallowance of dynamic shapes can affect caching, e.g., a change of
```
There was one failing unit test that exposed the following special-case. The test was expecting an exception to be thrown after changing this mode, but I was getting a cache hit (correctly according to the dispatch args), and therefore didn't raise. This is another thing that makes me question whether I've really found every case where behavior is subject to some external mode or global.

eellison replied:
That is pretty strange. Link to test ?

masnesral replied:
@eellison : `test_opcheck_opinfo_NumpyNMSCustomOp_cuda_float32`
Here I think: https://github.com/pytorch/pytorch/blob/main/test/test_custom_ops.py#L326-L330

masnesral replied:
Perhaps there's a more direct top-level setting to capture whether dynamic shapes are allowed or disallowed?

masnesral commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1414,6 +1524,215 @@ def __exit__(self, a, b, c):
+        elif isinstance(arg, torch.Tensor):
+            raise _BypassDispatchCache("non-fake tensor")
+        else:
+            # It's important to capture the type of the arg since, e.g., 1 and 1.0
```
I learned this the hard way. Annoying.

masnesral commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1414,6 +1524,215 @@ def __exit__(self, a, b, c):
+            raise _BypassDispatchCache("non-FakeTensor output")
+
+        # Avoid caching FakeTensors with constants attached since those
+        # can be invalidated.
```
@eellison I heard you might know something about the constant field that can be attached to the FakeTensor. Am I right that it would be challenging to support correctly with this FakeTensor cache?

eellison replied:
Hmm, it's at least more challenging, yea. How often our we getting cache misses because of it ? I think definitely in the first pr not needed. If it's a huge source of misses we can look into it.

masnesral replied:
Well on Chillee's gpt-fast at least, it doesn't show up:

hits = 22,435
misses = 523
bypasses = {
  'dynamic output shape': 8,
  'CompositeImplicitAutograd': 306,
  'non-FakeTensor output': 107,
  'lift': 60,
  'non-fake tensor': 35,
}

eellison replied:
Yea, let's not handle it for now

masnesral commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1414,6 +1524,215 @@ def __exit__(self, a, b, c):
+        if output.constant is not None:
+            raise _BypassDispatchCache("constant attribute")
+
+        # TODO: support caching sparse outputs?
```
wdyt?

bdhirsh replied:
I don't think sparse ops are very well supported today by FakeTensor, seems ok to skip for now

masnesral commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1437,12 +1756,6 @@ def dispatch(self, func, types, args=(), kwargs=None):
         elif func is torch.ops.aten.storage_offset.default:
             return int(args[0].storage_offset())
 
-        if log.getEffectiveLevel() <= logging.DEBUG:
```
Looks like I moved this below which I don't think was necessary. I'll move it back.

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1414,6 +1524,215 @@ def __exit__(self, a, b, c):
+    def _cached_dispatch_impl(
+        self,
+        func: OpOverload,
+        types: Tuple[Any],
+        args: Tuple[Any],
```
nit, I think you want `Tuple[Any..]` for both of these

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1414,6 +1524,215 @@ def __exit__(self, a, b, c):
+            FakeTensorMode.cache_hits += 1
+            output = self._output_from_cache_entry(entry, func, args)
+        else:
+            FakeTensorMode.cache_misses += 1
+            output = self._dispatch_impl(func, types, args, kwargs)
+            try:
+                self._create_cache_entry(key, args, output)
+            except _BypassDispatchCache as e:
+                FakeTensorMode.cache_bypasses[e.reason] += 1
```
I guess this is for something like non sym int inputs, sym int outputs

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1414,6 +1524,215 @@ def __exit__(self, a, b, c):
+        if func == aten._unsafe_view.default:
+            raise _BypassDispatchCache("unsafe view")
+
+        if func == aten.set_.source_Tensor:
+            raise _BypassDispatchCache("set source")
```
Can we add `set_` to `inplace_view` ? cc @bdhirsh

bdhirsh replied:
yep, I think we should

masnesral replied:
Cool. @@bdhirsh can I get a hint on how to do that. I thought it might be as simple as adding `tags: inplace_view` in native_functions.yaml, but I get an assertion that makes it look like that's not so simple:

`AssertionError: set_.source_Tensor is marked with tag: inplace_view. The codegen expects there to be a corresponding out-of-place view op with the name 'set_' and matching schema, but it didn't find one.`

bdhirsh replied:
ugh. you're gonna have to massage the codegen a bit, since the codegen makes some assumptions about ops with the `inplace_view` tag, but `set_()` is special. That comment sums it up: codegen expects that every inplace_view has a corresponding out-of-place view op. But `set_()` is a crazy op that swaps out storage, and doesn't have one.

bdhirsh replied:
It might be enough to add `and str(f.func.name.name) != "set_"` to here https://github.com/pytorch/pytorch/blob/main/torchgen/gen.py#L276. Can you give that a try and see if anything else breaks?

masnesral replied:
I'll remove this once we get https://github.com/pytorch/pytorch/pull/115769 working

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1414,6 +1524,215 @@ def __exit__(self, a, b, c):
+        # In order to handle storage aliasing, we need to establish the alias
+        # for any view op on a cache hit. But CompositeImplicitAutograd ops may
+        # or may not alias the input, so just punt on caching these.
+        if func.is_view and torch._C._dispatch_has_kernel_for_dispatch_key(
+            func.name(), torch._C.DispatchKey.CompositeImplicitAutograd
+        ):
+            raise _BypassDispatchCache("CompositeImplicitAutograd")
```
cc @bdhirsh, you were talking about an easier way of doing this right ?

bdhirsh replied:
the code above is the "easier way" I had in mind 😛 .

This code is basically querying the source of truth (the dispatcher state), to ask if there are any CompositeImplicitAutograd registrations

bdhirsh replied:
In theory we could cache on some CompositeImplicitAutograd ops, the problem is that we haven't actually done the work to find and test for the set of CompositeImplicitAutograd ops that are not cache friendly. So instead we can just decompose them all here (aka what we're doing right now)

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1414,6 +1524,215 @@ def __exit__(self, a, b, c):
+            self._translate_arg(kwargs),
+            # Capture the default_dtype mode since that can affect the output tensor,
+            # e.g., when operating on constant float values.
+            torch.get_default_dtype(),
```
good catch.. wonder if there's anything else we need to cache on globally 🤔 amp and grad_enabled should already be handled through torch_dispatch

masnesral replied:
AFAICT, the device is always available in the Tensor so we capture it as part of the TensorMetadata. Are there non-Tensor situations where the default device could affect the operation?

peterbell10 replied:
IIUC this code will happily cache tensor creation, where there isn't necessarily a tensor to take the device from. So default device should be relevant.

bdhirsh replied:
@peterbell10 beat me too it :) (matters for tensor creation)

peterbell10 replied:
Looks like `torch.backends.cuda.preferred_linalg_library()` can also change the output strides of linalg ops:
https://github.com/pytorch/pytorch/blob/afe6d272c69ae5671ca0df978be8fff7e8e4ed4e/aten/src/ATen/native/BatchLinearAlgebra.cpp#L682C62-L684

masnesral replied:
Ugh, how many things like this could be out there? For this one, would people recommend adding that to the list of items we use to compute the key or skip caching linalg ops? cc @eellison

eellison replied:
That is a good point @peterbell10. However, I'll note that it's probably pretty unlikely that dynamo guards on this currently. If someone were to change the setting, we'd end up in a hard error at runtime when inductor checks strides, before or after this pr.

cc @yanboliang, is there a list somewhere of the all the global torch settings we guard on ? Maybe we could factor that out so that the fake tenosr caching here shares the logic.

yanboliang replied:
https://github.com/pytorch/pytorch/pull/111299 is trying to added guard on dynamo config, but it has been reverted due to internal SEV. I think we can guard on global torch settings in a similar way.

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1414,6 +1524,215 @@ def __exit__(self, a, b, c):
+
+        # If this is an in-place op, we just record which input arg is aliased.
+        entry = None
+        for idx in range(len(args)):
+            if id(args[idx]) == id(output):
+                entry = _DispatchCacheEntry(None, idx)
+                break
+
```
I think we need to handle `kwargs` as well. And I wonder if there's any nested data structures that matter here..

masnesral replied:
@eellison Oh really? Do you know of any ops that alias one of the kwarg inputs that I can play with?

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1414,6 +1524,215 @@ def __exit__(self, a, b, c):
+        FakeTensorMode.cache[key] = entry
+
+    def _output_from_cache_entry(
+        self, entry: _DispatchCacheEntry, func: OpOverload, args: Tuple[Any]
```
Seems like we need to handle `kwargs` as well

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1414,6 +1524,215 @@ def __exit__(self, a, b, c):
+        torch._C._set_neg(empty, metadata.is_neg)
+
+        if func.is_view or metadata.storage_offset != 0:
+            if func.is_view:
+                # For view ops, the storage should be the same as the input.
+                tensor_args = [t for t in args if isinstance(t, torch.Tensor)]
+                assert len(tensor_args) == 1
+                storage = tensor_args[0].untyped_storage()
```
Hmm, we do we have this invariant? That if it's a view there is only one tensor and it's aliased. For instance, `index_put_` does not follow this. 

What if `index_put_ is called with `self` as a kwarg ? Seems like we need to handle kwargs here as well.

bdhirsh replied:
> For instance, index_put_ does not follow this.

I think this is actually true in practice today for ops that are `.is_view` (index_put_ doesn't register as a view op, since it mutates its input)
```
>>> torch.ops.aten.index_put_.default.is_view
False
```

I am a bit worried that there are edge cases we are missing for view support, but I can't think of any. Since e.g. a lot of the complexity for properly returning a view is around getting the autograd metadata correct (the `.grad_fn` and `.is_view` fields), and we don't have to worry about that here since we're below autograd

bdhirsh replied:
My takeaway (but elias can tell me if he disagrees) is:

(1) the code as written assumes that if `op.is_view` is true, then it must have a **single** tensor arg, guaranteed to be `args[0]`, which is the tensor arg that aliases the output tensor

(2) `is_view` only returns true for ops with "immutable aliasing", aka it will not return true for inplace ops like `.index_put_`, or `add_.Tensor`

(3) Therefore I think this is a safe assumption to make today - that assumption is the case for all view ops today that are a part of ATen

eellison reviewed (commented):
Looks great !!! I have only two main comments:

- Seems like kwargs are not fully handled
- Would be great to add a mode that will check correctness of cache hits versus re-running the op. Also don't have any tests - this would be a good way of testing. We can integrate into existing fake tensor tests.

masnesral commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1306,6 +1309,107 @@ def tolist(self):
+    in-place op and we need to alias the given input.
+    """
+
+    metadata: Optional[TensorMetadata] = None
```
Maybe I misunderstand, but I _think_ that's what I'm trying to capture with `arg_idx`. If `arg_idx` is set in the cache entry, then we alias the appropriate input arg when serving from the cache. If `metadata` is set in the cache entry instead, then we create a new FakeTensor from the metadata and serve that. So either there's another case I don't understand or my comment above is crap.

eellison replied:
Yea - nvm, that's what this is..

masnesral committed:
Update on "Initial implementation of FakeTensor caching"


Summary: Cache the result of FakeTensor dispatch and skip re-evaluation on cache hits.

Test Plan: New unit tests. Caching is enabled in this diff, so all existing tests exercise the cache as well.

[ghstack-poisoned]

bdhirsh commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1414,6 +1528,249 @@ def __exit__(self, a, b, c):
+            if not self.is_our_fake(arg):
+                raise _BypassDispatchCache("not our fake")
+            if arg._has_symbolic_sizes_strides:
+                raise _BypassDispatchCache("symbolic shape")
```
Is there a plan for dynamic shapes? (just support static for now and worry / think about dyamic shape plan later?)

masnesral replied:
> (just support static for now and worry / think about dyamic shape plan later?)

Yeah, that's the plan. Supporting dynamic shapes for the code cache was a bit tricky. Basically, we had to:
1) create a guards expression when we encounter dynamic shapes
2) save that expression with the cache entry
3) on a potential "hit", evaluate the expression in the current context to see if a "hit" is really a "hit"

I presume we'd have to take a similar approach to support dynamic shapes here? But I think it's trickier here because the performance of the cache lookup is more critical compared to the code cache.

bdhirsh reviewed (commented):
nice!

masnesral committed:
Update on "Initial implementation of FakeTensor caching"


Summary: Cache the result of FakeTensor dispatch and skip re-evaluation on cache hits.

Test Plan: New unit tests. Caching is enabled in this diff, so all existing tests exercise the cache as well.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

masnesral committed:
Update on "Initial implementation of FakeTensor caching"


Summary: Cache the result of FakeTensor dispatch and skip re-evaluation on cache hits.

Test Plan: New unit tests. Caching is enabled in this diff, so all existing tests exercise the cache as well.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1316,6 +1423,13 @@ def tolist(self):
+    cache_misses: int = 0
+    cache_bypasses = defaultdict(int)
+    cache_enabled: bool = FakeTensorConfig.cache
+    cache_crosscheck: bool = FakeTensorConfig.cache_crosscheck
```
What's the point of having `FakeTensorConfig` versions of these variables if they're only read at import time?

masnesral replied:
oops - Left over from an earlier impl where the settings weren't class variables in FakeTensorConfig. I'll fix!

peterbell10 commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1414,6 +1528,272 @@ def __exit__(self, a, b, c):
+            # The disallowance of dynamic shapes can affect caching, e.g., a change of
+            # mode can introduce a DynamicOutputShapeException where it wasn't seen on
+            # a previous instance of the same func and arg combination.
+            self.shape_env and self.shape_env.allow_dynamic_output_shape_ops,
```
Should you be checking the other shape_env config flags as well?

masnesral replied:
Sounds safer to do that.

masnesral committed:
Update on "Initial implementation of FakeTensor caching"


Summary: Cache the result of FakeTensor dispatch and skip re-evaluation on cache hits.

Test Plan: New unit tests. Caching is enabled in this diff, so all existing tests exercise the cache as well.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

masnesral commented on test/inductor/test_codecache.py:
```diff
@@ -132,12 +135,6 @@ def fn(x, y):
         self.assertEqual(counters["inductor"]["fxgraph_cache_miss"], 1)
         self.assertEqual(counters["inductor"]["fxgraph_cache_hit"], 1)
 
-        # But we expect different code if the tensors are aliased.
```
This was never correct and was passing by mistake. Enabling the FakeTensor cache exposes the mistake (by mistake ;) ).

eellison commented on test/test_ops.py:
```diff
@@ -85,6 +86,9 @@
 
 assert torch.get_default_dtype() == torch.float32
 
+FakeTensorConfig.cache_enabled = True
+FakeTensorConfig.cache_crosscheck_enabled = True
```
Maybe it makes more sense to just patch this within `TestFakeTensor` as well as `test_fake_tensor.py`

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1414,6 +1536,281 @@ def __exit__(self, a, b, c):
+            # The disallowance of dynamic shapes could introduce a
+            # DynamicOutputShapeException where it wasn't seen on a previous instance
+            # of the same op.
+            shape_env.allow_scalar_outputs if shape_env else None,
+            shape_env.allow_dynamic_output_shape_ops if shape_env else None,
+            shape_env.assume_static_by_default if shape_env else None,
+            shape_env.specialize_zero_one if shape_env else None,
+            shape_env.duck_shape if shape_env else None,
```
Can we factor this out to a dataclass like `ShapeEnvSettings` and add `xxx: add new settings to ShapeEnvSettings` comment in shape env `__init__` ? 

If we were to be extra cautious we could even commit a test which checks that all `boolean` fields in shape_env are tracked in the `ShapeEnvSettings`

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1031,8 +1033,22 @@ def should_allow_numbers_as_tensors(func: OpOverload):
+
 class FakeTensorConfig:
-    debug = os.environ.get("TORCH_FAKE_TENSOR_DEBUG", False)
+    debug = os.environ.get("TORCH_FAKE_TENSOR_DEBUG", "0") == "1"
+    cache_enabled = (
+        os.environ.get(
+            "TORCH_FAKE_TENSOR_DISPATCH_CACHE",
+            "0" if is_fbcode() else "1",
+        )
+        == "1"
+    )
+    cache_crosscheck_enabled = (
+        os.environ.get("TORCH_FAKE_TENSOR_DISPATCH_CACHE_CROSSCHECK", "0") == "1"
+    )
```
`debug` and `cache_crosscheck_enabled` make sense to put here but I think `cache_enabled` we should move to a more standardized place. `torch/_functorch/config.py` or `torch/_dynamo/config.py`. Just to standardize config options which affect runtime behavior. Debug options I think is okay to be a little bit more scattered.

eellison reviewed (approved):
🚀 🚀 🚀 🚢 🚢 🚢  I just had a few more comments but we should land this

zou3519 commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1414,6 +1536,281 @@ def __exit__(self, a, b, c):
+        if func == aten.set_.source_Tensor:
+            raise _BypassDispatchCache("set source")
+
+        if func in self.lift_fns:
+            raise _BypassDispatchCache("lift")
+
```
there's a helper func for this: torch._library.utils.is_builtin

masnesral committed:
Update on "Initial implementation of FakeTensor caching"


Summary: Cache the result of FakeTensor dispatch and skip re-evaluation on cache hits.

Test Plan: New unit tests. Caching is enabled in this diff, so all existing tests exercise the cache as well.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

masnesral committed:
Update on "Initial implementation of FakeTensor caching"


Summary: Cache the result of FakeTensor dispatch and skip re-evaluation on cache hits.

Test Plan: New unit tests. Caching is enabled in this diff, so all existing tests exercise the cache as well.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

masnesral committed:
Update on "Initial implementation of FakeTensor caching"


Summary: Cache the result of FakeTensor dispatch and skip re-evaluation on cache hits.

Test Plan: New unit tests. Caching is enabled in this diff, so all existing tests exercise the cache as well.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

masnesral committed:
Update on "Initial implementation of FakeTensor caching"


Summary: Cache the result of FakeTensor dispatch and skip re-evaluation on cache hits.

Test Plan: New unit tests. Caching is enabled in this diff, so all existing tests exercise the cache as well.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

masnesral committed:
Update on "Initial implementation of FakeTensor caching"


Summary: Cache the result of FakeTensor dispatch and skip re-evaluation on cache hits.

Test Plan: New unit tests. Caching is enabled in this diff, so all existing tests exercise the cache as well.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

masnesral committed:
Update on "Initial implementation of FakeTensor caching"


Summary: Cache the result of FakeTensor dispatch and skip re-evaluation on cache hits.

Test Plan: New unit tests. Caching is enabled in this diff, so all existing tests exercise the cache as well.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

masnesral committed:
Update on "Initial implementation of FakeTensor caching"


Summary: Cache the result of FakeTensor dispatch and skip re-evaluation on cache hits.

Test Plan: New unit tests. Caching is enabled in this diff, so all existing tests exercise the cache as well.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

masnesral committed:
Update on "Initial implementation of FakeTensor caching"


Summary: Cache the result of FakeTensor dispatch and skip re-evaluation on cache hits.

Test Plan: New unit tests. Caching is enabled in this diff, so all existing tests exercise the cache as well.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

masnesral commented:
@JackCaoG would you mind taking a look at the failing XLA test and advise? It's certainly possible that there's a problem with the impl that only surfaces on this test. I just don't know if we have a process / protocol for investigating failures in these xla tests. I'm asking around, but so far it's not clear to me if anyone internal is versed in running these tests locally.

JackCaoG commented:
Let me take a look, seems like a easy fix on our end.

JackCaoG commented:
Let's see if CI will pass in https://github.com/pytorch/xla/pull/6308

JackCaoG commented:
I think CI looks good on our end, you can update https://github.com/pytorch/pytorch/blob/main/.github/ci_commit_pins/xla.txt to 8a954a1b0c67eb5f844517a725c6532ceb07cb98 (last commit of my pr branch) and CI should start passing. Give me a ping when this pr is merged then I will take care of merging the XLA pr(it has to merge after this pr merges).

masnesral committed:
Update on "Initial implementation of FakeTensor caching"


Summary: Cache the result of FakeTensor dispatch and skip re-evaluation on cache hits.

Test Plan: New unit tests. Caching is enabled in this diff, so all existing tests exercise the cache as well.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

masnesral commented:
@pytorchbot merge

masnesral commented:
@masnesral has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D52841637).

huydhn commented:
@masnesral @JackCaoG Could you help update the XLA pinned commit as I think @JackCaoG has updated the the change on https://github.com/pytorch/xla/commit/d6dc1a01a4d8c9d3a4369fbc68a14192c60006f6?  XLA job is failing in trunk https://hud.pytorch.org/pytorch/pytorch/commit/208e64a9ba2abca63ea77b3b2dc5a2f19e4c2ac4 because it fails to checkout `reference is not a tree: 8a954a1b0c67eb5f844517a725c6532ceb07cb98`

JackCaoG commented:
Thanks, Let me double check

JackCaoG commented:
Hmm I think for some reason XLA started to delete the merged branch which cause XLA pin has to be a commit to the master. This will create various issues.. FYI @wonjoolee95 while I am trying to fix this..

JackCaoG commented:
Seems like this only happened to my branch, I updated https://github.com/pytorch/pytorch/pull/117708.

If https://github.com/pytorch/pytorch/pull/117434 merged first I think it will also fix the issue. In fact it might be better to let that pr merged...
